### DISCUSSION TITLE
yamlls does not properly receive capabilities

### DISCUSSION THREAD

**Discussion by Jomik** [LazyVim/LazyVim#568] [Q&A]
I started getting an error with `nvim-ufo` after having configured `yamlls`. I believe I configured it the exact same way as the json extras plugin. https://github.com/kevinhwang91/nvim-ufo/issues/127
The error seems to happen because the `textDocument/foldingRange` capability is not being sent to the server, despite it being set in my config.
It does seem to work for other language servers, but I am unsure how to exactly verify that they are sent, and I am not simply seeing good treesitter folds.
I do not use folding a lot yet.

folding.lua
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
    },
    {
      "kevinhwang91/nvim-ufo",
      dependencies = "kevinhwang91/promise-async",
      event = "BufReadPost",
      ---@class UfoConfig
      opts = {
        -- snip
      },
      keys = {
        -- snip
      },
    },
  },
}
```

yaml.lua
```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, { "yaml" })
    end,
  },
  {
    "neovim/nvim-lspconfig",
    dependencies = {
      "b0o/SchemaStore.nvim",
      version = false, -- last release is way too old
    },
    opts = {
      -- make sure mason installs the server
      ---@type lspconfig.options
      servers = {
        yamlls = {
          -- lazy-load schemastore when needed
          on_new_config = function(new_config)
            new_config.settings.yaml.schemas = new_config.settings.yaml.schemas or {}
            vim.list_extend(new_config.settings.yaml.schemas, require("schemastore").yaml.schemas())
          end,
          settings = {
            yaml = {
              format = { enable = true },
              hover = true,
              validate = true,
            },
          },
        },
      },
    },
  },
}
```

**Comment by Jomik** [comment]
I have updated the title to better reflect my issue.
Still struggling with this, really unsure what is going on.

**Comment by folke** [comment]
I noticed that as well. I think the problem is that yamls doesn't support folding, so you should not set the capability for that server

**Reply by Jomik** [reply]
Pretty certain it does support folding, see https://github.com/kevinhwang91/nvim-ufo/issues/72
I think the problem is that our client does not tell it that it has the folding capability, but `nvim-ufo` thinks we have told it, so it breaks.

**Reply by Jomik** [reply]
This seems to work, but I do not understand why I have to do this, when the above should already set these capabilities - and they seem to be set for other LS? Is there some good way for me to check?
```lua
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, { "yaml" })
    end,
  },
  {
    "neovim/nvim-lspconfig",
    dependencies = {
      "b0o/SchemaStore.nvim",
      version = false, -- last release is way too old
    },
    opts = {
      -- make sure mason installs the server
      ---@type lspconfig.options
      servers = {
        yamlls = {
          capabilities = {
            textDocument = {
              foldingRange = {
                dynamicRegistration = false,
                lineFoldingOnly = true,
              },
            },
          },
          -- lazy-load schemastore when needed
          on_new_config = function(new_config)
            new_config.settings.yaml.schemas = new_config.settings.yaml.schemas or {}
            vim.list_extend(new_config.settings.yaml.schemas, require("schemastore").yaml.schemas())
          end,
          settings = {
            yaml = {
              format = { enable = true },
              hover = true,
              validate = true,
            },
          },
        },
      },
    },
  },
}
```

**Reply by folke** [reply]
Where did you see lspconfig opts.capabilities being used?

**Reply by folke** [reply]
I'm adding support for opts.capabilities. Just a second :)

**Reply by folke** [reply]
Just added support for setting global capabilities and tested it with yamls. So your existing config should work now :)

Will create a new release in a bit.

**Reply by Jomik** [reply]
I have no clue! That is so weird, did not even think to check source code.. Awesome that you add support though! Think it makes perfect sense, as the client should always have the same capabilities, so should be little need to set it per server 👍 

Amazing as always!

**Reply by Jomik** [reply]
This should work on `2.5.0` right? It does not seem to 🤔 
https://github.com/Jomik/dotfiles/blob/9e428b25c65a3f89d634681a2227a1fd20f41d1a/.config/nvim/lua/plugins/folding.lua#L33-L42

**Reply by folke** [reply]
Yep, it does work on 2.5.0. Tested it before and after my changes. Works perfectly fine.

I do see that you just copied the jsonls settings to yamlls? That's a totally different server. I just have this for yamlls:

```lua
        yamlls = {
          settings = {
            yaml = {
              keyOrdering = false,
            },
          },
        },
```

And for nvim-ufo:

```lua
return {

  -- add folding range to capabilities
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
    },
  },

  -- add nvim-ufo
  {
    "kevinhwang91/nvim-ufo",
    dependencies = "kevinhwang91/promise-async",
    event = "BufReadPost",
    opts = {},

    init = function()
      -- Using ufo provider need remap `zR` and `zM`. If Neovim is 0.6.1, remap yourself
      vim.keymap.set("n", "zR", function()
        require("ufo").openAllFolds()
      end)
      vim.keymap.set("n", "zM", function()
        require("ufo").closeAllFolds()
      end)
    end,
  },
}
```

**Reply by Jomik** [reply]
> I do see that you just copied the jsonls settings to yamlls? That's a totally different server. I just have this for yamlls:
> 
> ```lua
>         yamlls = {
>           settings = {
>             yaml = {
>               keyOrdering = false,
>             },
>           },
>         },
> ```

They are similar [yaml-language-server#language-server-settings](https://github.com/redhat-developer/yaml-language-server#language-server-settings)
I will play around with it more tomorrow, must be something on my end 👍

**Reply by Jomik** [reply]
I cannot seem to figure it out. If I add the capability on `yamlls` it works just fine, but if I have the global one as above, it does not seem to work.
I have tried stripping the configs to be exactly what you provided.
LazyVim is now version 2.6.0.

**Comment by cjosue15** [comment]
HI i tried to follow your suggestion @folke in #448

```lua
return {

  -- add folding range to capabilities
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
    },
  },

  -- add nvim-ufo
  {
    "kevinhwang91/nvim-ufo",
    dependencies = "kevinhwang91/promise-async",
    event = "BufReadPost",
    opts = {},

    init = function()
      -- Using ufo provider need remap `zR` and `zM`. If Neovim is 0.6.1, remap yourself
      vim.keymap.set("n", "zR", function()
        require("ufo").openAllFolds()
      end)
      vim.keymap.set("n", "zM", function()
        require("ufo").closeAllFolds()
      end)
    end,
  },
}
```
But it seems there are some issues with neo-tree and the buffer itself, this is how it looks my nvim config when i open it :( and i can't open the folders in neotree

![image](https://user-images.githubusercontent.com/51803643/233494579-b7650db7-e29d-49fd-b789-bb54e8ac09c0.png)

So i decided to follow another config that i found here https://github.com/kevinhwang91/nvim-ufo/issues/117#issuecomment-1451569830

The result of my config is this 

```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
    },
  },
  {
    "kevinhwang91/nvim-ufo",
    dependencies = { "kevinhwang91/promise-async" },
    event = "BufRead",
    keys = {
      {
        "zR",
        function()
          require("ufo").openAllFolds()
        end,
      },
      {
        "zM",
        function()
          require("ufo").closeAllFolds()
        end,
      },
      {
        "K",
        function()
          local winid = require("ufo").peekFoldedLinesUnderCursor()
          if not winid then
            vim.lsp.buf.hover()
          end
        end,
      },
    },
    config = function()
      vim.o.fillchars = [[eob: ,fold: ,foldopen:,foldsep: ,foldclose:]]
      vim.o.foldcolumn = "1"
      vim.o.foldlevel = 99
      vim.o.foldlevelstart = 99
      vim.o.foldenable = true
      require("ufo").setup()
    end,
  },
}
```

this config looks good with lua files and scss files but i'm not sure what happend with typescript and html files, sometimes works and sometimes not works

![image](https://user-images.githubusercontent.com/51803643/233494990-0f46fbe7-6b9e-4a80-ba61-751b9a6b8a7d.png)

i tried as well with tha last config of lazy vim 2.6.0 but it not works :(

**Reply by dpetka2001** [reply]
I just used @Jomik dotfiles for [folding](https://github.com/Jomik/dotfiles/blob/master/.config/nvim/lua/plugins/folding.lua) and to resolve the issue with neo-tree i added the following autocmd 
```lua
-- Disable Ufo for neo-tree
autocmd("Filetype", {
  pattern = "neo-tree",
  callback = function()
    vim.cmd("UfoDetach")
  end,
  group = "mygroup",
  desc = "Disable Ufo for neo-tree",
})
```
Seems to be working fine on lua, html,js and css files for the time being. Don't use any typescript to test on.

**Comment by cjosue15** [comment]
Hi, I was surprised that my problem is related to "angular LSP", because angularls has its own folding so this is the reason why it conflicts and sometimes the folding worked and sometimes not.

So I decided to disable the property below, in order to get all the features of the angular language server except for the folding.

![image](https://user-images.githubusercontent.com/51803643/233691228-9956e00f-ecc6-403a-9f70-7d3cdf56b83e.png)

to check the status of this, you have to run this in cmd of nvim

```lua vim.lsp.get_active_clients()[index].server_capabilities```

What is index? In my case when i check LspInfo my angularls was in the position 3, so you have to check that as well

![image](https://user-images.githubusercontent.com/51803643/233690107-367d2f0b-54de-4790-97dc-5d41420628c0.png)

Finally i added this line of codes, to disabled the folding of angularls in my lazyVim local project, and it works!

```lua
        if client.name == "angularls" then
          client.server_capabilities.documentRangeFormattingProvider = false
          client.server_capabilities.foldingRangeProvider = false
        end
```

![image](https://user-images.githubusercontent.com/51803643/233691051-69de9cfd-80bd-4293-916e-8f4527dd0db5.png)

so @folke , I would like to know if you can add support for extending the on_attach funcition, something like the following :), sorry I am very new with nvim and lazy and not sure if this is a nice solution, or the solution already exists.

```lua
return {
  -- add folding range to capabilities
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
      on_attach = function(client, buffer)
        if client.name == "angularls" then
          client.server_capabilities.documentRangeFormattingProvider = false
          client.server_capabilities.foldingRangeProvider = false
        end
      end,
    },
  },
}
```

**Reply by cjosue15** [reply]
@Jomik maybe it can help you! with lsp yaml!

**Reply by folke** [reply]
You should **never** change LazyVim files or any plugins files for that matter.
You can simply add that `on_attach` to your `/lua/config/autocmds.lua` file

**Reply by folke** [reply]
However, even easier is to use the `setup` functionality. See example for clangd:

```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      diagnostics = { virtual_text = { prefix = "icons" } },
      setup = {
        clangd = function(_, opts)
          opts.capabilities.offsetEncoding = { "utf-16" }
        end,
      },
    },
  },

```

**Reply by dpetka2001** [reply]
@folke I've always had this question, but would you mind explaining what the difference is between `servers.server_name.settings` and `setup.server_name`? When should someone prefer the one over the other?

**Reply by cjosue15** [reply]
I know @folke but at the time I didn't know there was an option to add config to the lsp angular server, but after reading the lazyVim doc I found a solution without modifying their repo.

For me this is my solution, as i said maybe @Jomik you have a probem with the lsp of yaml like me with angular.

`/lua/plugins/lsp.lua`

```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      capabilities = {
        textDocument = {
          foldingRange = {
            dynamicRegistration = false,
            lineFoldingOnly = true,
          },
        },
      },
      setup = {
        angularls = function(_, opts)
          opts.on_attach = function(client, _)
            client.server_capabilities.documentRangeFormattingProvider = false
            client.server_capabilities.foldingRangeProvider = false
          end
        end,
      },
    },
  },
}

```
